R/colocalization_heatmap.R

```diff
@@ -88,31 +223,109 @@ ColocalizationHeatmap <- function(
       distinct()
   }
 
-  # Pivot results
-  plot_data <-
-    plot_data %>%
-    pivot_wider(names_from = "marker_2", values_from = "value", values_fill = 0) %>%
-    column_to_rownames("marker_1")
+  # Set range for heatmap legend
+  if (is.null(legend_range)) {
+    legend_range <- max(abs(plot_data %>% pull(all_of(value_col)))) * c(-1, 1)
+  }
 
-  if (return_plot_data) {
-    return(plot_data)
+  # Cap values to legend range
+  plot_data <- plot_data %>%
+    mutate(across(all_of(value_col), ~ case_when(
+      .x < legend_range[1] ~ legend_range[1],
+      .x > legend_range[2] ~ legend_range[2],
+      TRUE ~ .x
+    )))
+
+  # Transform size col
+  if (!is.null(size_col_transform)) {
+    stopifnot(
+      "size_col_transform must be a function" =
+        is.function(size_col_transform)
+    )
+    plot_data <- plot_data %>%
+      mutate(!!sym(size_col) := size_col_transform(!!sym(size_col)))
+    size_col_label <- paste0(size_col, "_transformed")
+  } else {
+    size_col_label <- size_col
   }
 
-  # Set range for heatmap legend
-  if (is.null(legend_range)) {
-    legend_range <- max(abs(plot_data)) * c(-1, 1)
+  # Cast long table to a wide matrix
+  # This is necessary for pheatmap (tiles) but not for the ggplot based version (dots)
+  # However, we also do it conditionally for dots if clustering is enabled to get the
+  # correct order of rows and columns
+  if (type == "tiles" || (cluster_rows || cluster_cols)) {
+    plot_data_wide <-
+      plot_data %>%
+      {
+        if (!is.null(size_col) && (size_col %in% names(.))) {
+          select(., -all_of(size_col))
+        } else {
+          .
+        }
+      } %>%
+      pivot_wider(names_from = "marker_2", values_from = all_of(value_col), values_fill = 0) %>%
+      column_to_rownames("marker_1") %>%
+      as.matrix()
+
+    # Cluster rows for the dots method
+    if (cluster_rows && (type == "dots")) {
+      rows_clust <- dist(plot_data_wide, method = clustering_distance_rows) %>%
+        hclust(method = clustering_method)
+      plot_data <- plot_data %>%
+        mutate(marker_1 = factor(marker_1, levels = rows_clust$labels))
```

Comment:
I think labels will not be in the clustered order, but alphabetic. As far as I know, the `order` is a separate element of the `hclust` output so you can use `with(rows_clust, labels[order]))` to get the levels in order:
```suggestion
        mutate(marker_1 = factor(marker_1, levels = with(rows_clust, labels[order])))
```

Reply:
Perfect, thanks!